src/dispatch/par_seq.rs

```diff
@@ -0,0 +1,406 @@
+            },
+            "Tried to add system with conflicting reads / writes"
+        );
+
+        Par {
+            head: Par {
+                head: self.head,
+                tail: sys,
+            },
+            tail: Nil,
+        }
+    }
+}
+
+/// A dispatcher intended to be used with
+/// `Par` and `Seq`  structures.
+///
+/// This is more flexible and performant than `Dispatcher`,
+/// however, you have to check conflicts yourself.
+/// That means you cannot run two systems in parallel
+/// which write to the same resource; if you'd do that,
+/// one of the systems will panic while trying to fetch
+/// the `SystemData`.
+///
+/// ## Thread-local systems
+///
+/// This dispatcher also allows more freedom
+/// for thread-local systems; you cannot execute wherever you want,
+/// just not in parallel with other systems (putting one inside
+/// `par!` will give you a compile-time error saying the `Send` requirement
+/// is unmet).
+///
+/// ## Examples
+///
+/// ```
+/// # extern crate rayon;
+/// #[macro_use(par, seq)]
+/// extern crate shred;
+///
+/// # use rayon::ThreadPool;
+/// #
+/// # use shred::{ParSeq, Resources, System};
+/// #
+/// # macro_rules! impl_sys {
+/// #     ($( $id:ident )*) => {
+/// #         $(
+/// #             impl<'a> ::shred::System<'a> for $id {
+/// #                 type SystemData = ();
+/// #                 fn run(&mut self, _: Self::SystemData) {}
+/// #             }
+/// #         )*
+/// #     };
+/// # }
+/// #
+/// # struct SysA;
+/// # struct SysB;
+/// # struct SysC;
+/// # struct SysD;
+/// # struct SysWithLifetime<'a>(&'a u8);
+/// # struct SysLocal(*const u8);
+/// #
+/// # impl_sys!(SysA SysB SysC SysD SysLocal);
+/// #
+/// # impl<'a, 'b> System<'a> for SysWithLifetime<'b> {
+/// #     type SystemData = ();
+/// #
+/// #     fn run(&mut self, _: Self::SystemData) {}
+/// # }
+///
+/// # fn main() {
+/// # #![cfg_attr(rustfmt, rustfmt_skip)]
+/// #
+/// # let pool = ThreadPool::new(Default::default()).unwrap();
+/// #
+/// # let mut res = Resources::new();
+/// let x = 5u8;
+///
+/// let mut dispatcher = ParSeq::new(
+///     seq![
+///         par![
+///             SysA,
+///             SysWithLifetime(&x),
+///             seq![
+///                 SysC,
+///                 SysD,
+///             ],
+///         ],
+///         SysB,
+///         SysLocal(&x as *const u8),
+///     ],
+///     &pool,
+/// );
+///
+/// dispatcher.dispatch(&mut res);
+/// # }
+/// ```
+pub struct ParSeq<P, T> {
+    run: T,
+    pool: P,
+}
+
+impl<P, T> ParSeq<P, T>
+where
+    P: Borrow<ThreadPool>,
+    T: for<'a> RunWithPool<'a>,
+{
+    /// Creates a new `ParSeq` dispatcher.
+    /// `run` is usually created by using the `par!` / `seq!`
+    /// macros.
+    pub fn new(run: T, pool: P) -> Self {
+        ParSeq { run, pool }
+    }
+
+    /// Dispatches the systems using `res`.
+    /// This involves zero virtual cost.
```

Comment:
s/involves zero virtual cost/doesn't do any virtual function calls/

Reply:
If the cost is only virtual should we avoid it? 😄

Reply:
I feel that you *make* virtual function calls, but I can't tell for sure.